cmd/quadlet/main.go

```diff
@@ -681,4 +707,5 @@ func init() {
 	flag.BoolVar(&isUserFlag, "user", false, "Run as systemd user")
 	flag.BoolVar(&dryRunFlag, "dryrun", false, "Run in dryrun mode printing debug information")
 	flag.BoolVar(&versionFlag, "version", false, "Print version information and exit")
+	flag.StringVar(&listImagesFlag, "list-images", "", "Write a list of all images being references in .image Quadlets to specified file")
```

Comment:
Given that we're likely to want to add something like a new flag to opt-in to pinning like
`Bootc=bound` or `Preload=true` or `Pin=true` (however we call it)

I think we'll need to either:

- Add some sort of `--matches=Pin=true` flag to this list operation
- Just return the entire `.image` file?

I think the second part is important: Consider here that IMO is quite valid: let's say I want to use a custom pull secret just for some images. In that general case today, we need the `AuthFile` flag, etc.

This type of thing implies actually that what we will need as a next step is a way to pass the *output* of this tool (probably a full `.image` file) into something like `podman pull --from-systemd-image` or something`?

Reply:
Podman should not parse quadlet files. That is breaking the well build separation we have. And quite frankly it cannot work as systemd untis may contain systemd specifiers that must be expanded first and this is not logic we should try to mirror because we will fail to match it 1 to 1.
And even if you try to parse that on bootc you will have the same issue, the AuthFile=%h/somefile you have to know what %h is. And then you also have to deal with env var expansions and so on.

As such I fail to see how we can generate the command exactly the way systemd would execute it.

Reply:
Specifiers are a complex topic indeed. I think `%h` (homedir) specifically is out of scope for bootc's use case; we should not be interacting with user containers or home directories.

Would people use e.g. `%a` (architecture) or `%l` (hostname) or other things like that in these `.image` files? Maybe...but, honestly I'd also be OK if we just errored out if you used those things for `Pin=true` `.image` files today. I think we'd just need to handle "detect unquoted % in input and error out".

Reply:
So to combine these then, what we could output instead is say JSON, with a list of `podman pull ...` commands that we would run, and again error out if we find specifiers.

Reply:
Although it's a bit messy because if we also want to do pinning via `podman create --rm <image>`...wait, is there no variant of `create` that is explicitly ephemeral i.e. its state is tracked in `/run`? I guess that's not fatal, but it'd be quite nice to have as it means the core logic would just be "on boot, re-pin all images we need to pin" instead of that plus an extra step of "also GC any pin containers that aren't needed anymore".

Reply:
I want to extract some requirements from the conversation:
* We want to pre-pull images after an `bootc upgrade`
* Not all `.image` files may need to be pre-pulled, so we need some kind of opt-in mechanism
* The `.image` files may use custom pull secrets, so Quadlet should ideally have generated units for the file with a fully expanded `podman pull`
* Respecting the separation of concerns, Quadlet should not leave is purpose of a systemd-generator

Reply:
> We want to pre-pull images after an bootc upgrade

It's subtler than that I think. We want to "logically bind", and that means that bootc *must not queue a new bootloader entry* unless the logically bound images have been pulled and are known to be present.

A whole big point of bootc (and ostree before it, and image based systems in general) is having "transactional updates" where the system is known to either be in state A or B.

If we did the naive thing of:

- bootc upgrade (queuing a new bootloader entry)
- pull images

Then some tool doing a `reboot` in between these two phases would create a state like "B, but without app images" which is not desired.

This subtlety is why this is hard to implement without changes to both bootc and podman today.